### PR TITLE
Fix handlers

### PR DESCRIPTION
**Pull request by Jeasmine**
Fix Android handler 

* Add in-app message handler listener
* Fix grouped notification parsing, not add to JSON if is an empty array
* Fix action id, make it optional
* Add notification open handler listener

Fix iOS Notification open handler

* Add open handler listener
* Add cold start support --> Not working

On iOS notification, the open handler is not being called when the app is swiped away. 

## 
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375)

### CODE REVIEW

**Review by jkasten2 — COMMENTED** [review]
The comments above are something I think we should do before the final release. But ok to merge this PR as-is now to not block a Beta release.

> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-:-MXrtnkF7xPA2z7mRAKl:b5c17gj)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @emawby, @Jeasmine, and @rgomezp)

**Review by Jeasmine — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @emawby, @Jeasmine, @jkasten2, and @rgomezp)

___
*[ios/Classes/OneSignalPlugin.m, line 332 at r1](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-MXnJMlbDhhMFS9rTKhW:-MXsGQmx0JhqLexbH_ya:b-54uq67) ([raw file](https://github.com/onesignal/onesignal-flutter-sdk/blob/ce0c4b00545dc395de1d8013f89236810a2b24d0/ios/Classes/OneSignalPlugin.m#L332)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

The wrapper SDK shouldn't have to handle the cold start as the main SDK already handles this.
https://github.com/OneSignal/OneSignal-iOS-SDK/pull/826

For this to work correctly however we should call this code when the flutter developer setups the handler, not in `addObservers` above.
```
    [OneSignal setNotificationOpenedHandler:^(OSNotificationOpenedResult * _Nonnull result) {
       [OneSignalPlugin.sharedInstance handleNotificationOpened:result];
    }];

``` 
</blockquote></details>

Nice catch, 100% right, will remove this

**Review by Jeasmine — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @emawby, @jkasten2, and @rgomezp)

___
*[android/src/main/java/com/onesignal/flutter/OneSignalPlugin.java, line 46 at r1](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-MXnKSToBFmJCnYqp11d:-MXsKCC68lZ_-Ba7GFm9:b-896fix) ([raw file](https://github.com/onesignal/onesignal-flutter-sdk/blob/ce0c4b00545dc395de1d8013f89236810a2b24d0/android/src/main/java/com/onesignal/flutter/OneSignalPlugin.java#L46)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

See my iOS note about cold start Notification handling and match this with Android.
</blockquote></details>

Done.
___
*[ios/Classes/OneSignalPlugin.m, line 332 at r1](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-MXnJMlbDhhMFS9rTKhW:-MXsKCzi6x0ErbrrM2eJ:b-74hu5j) ([raw file](https://github.com/onesignal/onesignal-flutter-sdk/blob/ce0c4b00545dc395de1d8013f89236810a2b24d0/ios/Classes/OneSignalPlugin.m#L332)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

Nice catch, 100% right, will remove this
</blockquote></details>

After the change, on iOS, it still doesn't trigger the notification open handler when the app is swiped again under cold start. I also tested it while in the foreground.

Run application without setting the notification open handler.
Send a notification, and the device receives the notification
Open notification
Set open handler
Expected behavior, notification open handler callback is being called
Actual behavior only "Notification will show in foreground handler set successfully" msg appears on the log

**Review by Jeasmine — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @emawby, @jkasten2, and @rgomezp)

___
*[ios/Classes/OneSignalPlugin.m, line 332 at r1](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-MXnJMlbDhhMFS9rTKhW:-MXt2AnTD62FflyTFgRQ:bg6h0eh) ([raw file](https://github.com/onesignal/onesignal-flutter-sdk/blob/ce0c4b00545dc395de1d8013f89236810a2b24d0/ios/Classes/OneSignalPlugin.m#L332)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

After the change, on iOS, it still doesn't trigger the notification open handler when the app is swiped again under cold start. I also tested it while in the foreground.

Run application without setting the notification open handler.
Send a notification, and the device receives the notification
Open notification
Set open handler
Expected behavior, notification open handler callback is being called
Actual behavior only "Notification will show in foreground handler set successfully" msg appears on the log
</blockquote></details>

This scenario does work on Android

**Review by jkasten2 — CHANGES_REQUESTED** [review]
I am pretty sure the reason why iOS Notification open cold starts are not working as `OneSignal.initWithLaunchOptions` is being called too late.

In the 3.0.0 branch it is being called from `setAppId`:
https://github.com/OneSignal/OneSignal-Flutter-SDK/blob/bfadde23bd7d0f26f732776468ad1c014a57bd22/ios/Classes/OneSignalPlugin.m#L156

We used to call OneSignal init from `registerWithRegistrar` in the 2.x.x SDK:
https://github.com/OneSignal/OneSignal-Flutter-SDK/blob/2.6.3/ios/Classes/OneSignalPlugin.m#L85

> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-:-MXvB6ZO56p-TqkOCmTQ:b-o4dfud)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @emawby, @jkasten2, and @rgomezp)

**Review by jkasten2 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375)* status: 1 of 4 files reviewed, 2 unresolved discussions (waiting on @emawby, @Jeasmine, @jkasten2, and @rgomezp)

___
*[android/src/main/java/com/onesignal/flutter/OneSignalPlugin.java, line 46 at r1](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-MXnKSToBFmJCnYqp11d:-MXvC3SpAHBJjamjy1Mt:bdpuf89) ([raw file](https://github.com/onesignal/onesignal-flutter-sdk/blob/ce0c4b00545dc395de1d8013f89236810a2b24d0/android/src/main/java/com/onesignal/flutter/OneSignalPlugin.java#L46)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I see the code changes where the use is now removed for `coldStartNotificationResult`. Now we just need to delete this unused definition :smile:

**Comment by MeAtPros** [comment]
@Jeasmine Can you fix this error quickly? My client is waiting for the application update and we are unable to release the correct version due to this problem.

**Review by Jeasmine — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375)* status: 1 of 4 files reviewed, 2 unresolved discussions (waiting on @emawby, @jkasten2, and @rgomezp)

___
*[android/src/main/java/com/onesignal/flutter/OneSignalPlugin.java, line 46 at r1](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-MXnKSToBFmJCnYqp11d:-MYBRx4-Bk71NMbvsSq3:b-896fix) ([raw file](https://github.com/onesignal/onesignal-flutter-sdk/blob/ce0c4b00545dc395de1d8013f89236810a2b24d0/android/src/main/java/com/onesignal/flutter/OneSignalPlugin.java#L46)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

I see the code changes where the use is now removed for `coldStartNotificationResult`. Now we just need to delete this unused definition :smile: 
</blockquote></details>

Done.
___
*[ios/Classes/OneSignalPlugin.m, line 332 at r1](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-MXnJMlbDhhMFS9rTKhW:-MYBSL0rEqY0wMVRnY_B:b-vhrwcr) ([raw file](https://github.com/onesignal/onesignal-flutter-sdk/blob/ce0c4b00545dc395de1d8013f89236810a2b24d0/ios/Classes/OneSignalPlugin.m#L332)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

This scenario does work on Android
</blockquote></details>

Done. Fixed by moving initWithLaunchOptions

**Review by jkasten2 — APPROVED** [review]
Just one logging message to clean up. Feel free to merge in after fixing without another approval.

> Reviewed 1 of 4 files at r1, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-:-MYC5Cx7-hgpDxuMp21Q:bk1grw5)* status: all files reviewed, 1 unresolved discussion (waiting on @emawby, @Jeasmine, and @rgomezp)

___
*[ios/Classes/OneSignalPlugin.m, line 295 at r4](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-MYC511K1OJWKjKBiKr6:-MYC511K1OJWKjKBiKr7:burxfj7) ([raw file](https://github.com/onesignal/onesignal-flutter-sdk/blob/674a57116ebb13f405a8e6b11b4fe6db69d1c02d/ios/Classes/OneSignalPlugin.m#L295)):*
> ```objc
> - (void)initNotificationOpenedHandlerParams {
>     [OneSignal setNotificationOpenedHandler:^(OSNotificationOpenedResult * _Nonnull result) {
>         [OneSignal onesignalLog:ONE_S_LL_VERBOSE message:@"setNotificationOpenedHandler called from addObservers"];
> ```

Logging isn't correct. Replace "addObservers" with "initNotificationOpenedHandlerParams".

**Review by Jeasmine — COMMENTED** [review]
That log is not needed anymore since it is used to track open handler on cold start. But that is already fixed. I will remove the log

> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-flutter-sdk/375#-:-MYFjej46zU9AM6iXimt:b-nmlyog)* status: all files reviewed, 1 unresolved discussion (waiting on @emawby, @Jeasmine, and @rgomezp)

**Pull request merged**
